File: functional_thinking_examples/scala/number_classifier/NumberClassifier.scala
package main.scala

// BEGIN Number_Classifier_Scala
object NumberClassifier {
  def isFactor(factor: Int, number: Int) =
    number % factor == 0

  def factors(number: Int) =
    (1 to number) filter (isFactor(_, number))

  def sum(factors : Seq[Int]) =
    factors.foldLeft(0)(_ + _)

  def isPerfect(number: Int) =
    sum(factors(number)) - number == number

  def isAbundant(number: Int) =
    sum(factors(number)) - number > number

  def isDeficient(number: Int) =
    sum(factors(number)) - number < number
}
// END Number_Classifier_Scala

File: functional_thinking_examples/java/number_classifier/NumberClassifier.java
package nealford.ft.numberclassification;

// BEGIN functional_number_classifier
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class NumberClassifier {

    public static boolean isFactor(final int candidate, final int number) {   //<1> 
        return number % candidate == 0;                                    
    }                                                                      
                                                                           
    public static Set<Integer> factors(final int number) {                    //<2>
        Set<Integer> factors = new HashSet<>();                            
        factors.add(1);                                                    
        factors.add(number);                                               
        for (int i = 2; i < number; i++)                                   
            if (isFactor(i, number))                                       
                factors.add(i);                                            
        return factors;                                                    
    }                                                                      
                                                                           
    public static int aliquotSum(final Collection<Integer> factors) {
        int sum = 0;
        int targetNumber = Collections.max(factors);
        for (int n : factors) {                                               //<3>
            sum += n;
        }
        return sum - targetNumber;
    }
                                                                           
    public static boolean isPerfect(final int number) {                    
        return aliquotSum(factors(number)) == number;
    }                                                                      
                                                                              //<4>
    public static boolean isAbundant(final int number) {
        return aliquotSum(factors(number)) > number;
    }

    public static boolean isDeficient(final int number) {
        return aliquotSum(factors(number)) < number;
    }
}
// END functional_number_classifier
